File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.py
class Solution:
    def minSwaps(self, nums: List[int]) -> int:
        # count the total number of ones
        no_of_ones = nums.count(1)
        window_ones = 0
        max_window_ones = 0
        N = len(nums)
        i = 0
        j = 0

        while (j < 2 * N):
            if (nums[j % N] == 1):
                window_ones += 1

            if (j - i + 1 > no_of_ones):
                if (nums[i % N] == 1):
                    window_ones -= 1
                i += 1

            max_window_ones = max(max_window_ones, window_ones) 

            j += 1

        return no_of_ones - max_window_ones  

File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.js
var minSwaps = function(nums) {
    let countOfOnes = 0
    let k = nums.length

    for (let num of nums) {
        countOfOnes += num
    }
    
    let i = 0
    let j = 0
    let maxWindowOnes = 0
    let windowOnes = 0

    while (j < nums.length * 2) {
        if (nums[j % k]) windowOnes += 1

        while (j - i + 1 > countOfOnes) {
            if (nums[i % k]) windowOnes -= 1
            i += 1
        }
        maxWindowOnes = Math.max(maxWindowOnes, windowOnes)
        j += 1 
    }
    return countOfOnes - maxWindowOnes

};